File: android/src/main/kotlin/ai/synheart/session/WatchSessionRelay.kt
package ai.synheart.session

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.google.android.gms.common.GoogleApiAvailability
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.wearable.MessageClient
import com.google.android.gms.wearable.MessageEvent
import com.google.android.gms.wearable.Wearable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import org.json.JSONObject

/**
 * Phone-side relay that bridges session commands and events between the
 * Flutter plugin and a companion Wear OS watch app via the Wearable Data Layer.
 *
 * Mirrors the iOS [WatchSessionRelay.swift] behaviour:
 *   1. Plugin calls [startSession] → relay sends command to watch via [MessageClient].
 *   2. Watch engine emits events → relay receives them in [onMessageReceived]
 *      and forwards to the stored [eventCallback].
 *   3. On terminal events (`session_summary` / `session_error`) the relay cleans up.
 */
class WatchSessionRelay(private val context: Context) : MessageClient.OnMessageReceivedListener {

    companion object {
        private const val TAG = "WatchSessionRelay"
        const val EVENT_PATH = "/synheart/session/event"
        const val COMMAND_PATH = "/synheart/session/command"
    }

    private val messageClient: MessageClient = Wearable.getMessageClient(context)
    private val nodeClient = Wearable.getNodeClient(context)
    private val mainHandler = Handler(Looper.getMainLooper())
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

    private var eventCallback: ((Map<String, Any?>) -> Unit)? = null
    private var activeSessionId: String? = null

    init {
        messageClient.addListener(this)
    }

    // -- Status ---------------------------------------------------------------

    /**
     * Full connectivity snapshot for the Dart layer.
     * Matches the iOS format: { supported, reachable, paired, installed }.
     *
     * On Android, [paired] and [installed] are true when at least one connected
     * node exists — Wear OS does not expose these as separate properties.
     */
    fun getStatus(callback: (Map<String, Any>) -> Unit) {
        val supported = isPlayServicesAvailable()
        if (!supported) {
            callback(mapOf(
                "supported" to false,
                "reachable" to false,
                "paired" to false,
                "installed" to false
            ))
            return
        }

        scope.launch {
            val status = try {
                val nodes = nodeClient.connectedNodes.await()
                val hasNode = nodes.isNotEmpty()
                mapOf<String, Any>(
                    "supported" to true,
                    "reachable" to hasNode,
                    "paired" to hasNode,
                    "installed" to hasNode
                )
            } catch (e: Exception) {
                Log.w(TAG, "getStatus failed: ${e.message}")
                mapOf<String, Any>(
                    "supported" to true,
                    "reachable" to false,
                    "paired" to false,
                    "installed" to false
                )
            }

            mainHandler.post { callback(status) }
        }
    }

    /** Whether at least one Wear OS node is currently connected. */
    fun checkReachable(callback: (Boolean) -> Unit) {
        if (!isPlayServicesAvailable()) {
            Log.w(TAG, "Google Play Services NOT available — watch relay disabled")
            callback(false)
            return
        }

        scope.launch {
            val reachable = try {
                val nodes = nodeClient.connectedNodes.await()
                Log.d(TAG, "connectedNodes: ${nodes.size} node(s) found")
                for (node in nodes) {
                    Log.d(TAG, "  node: id=${node.id} name=${node.displayName} nearby=${node.isNearby}")
                }
                nodes.isNotEmpty()
            } catch (e: Exception) {
                Log.w(TAG, "connectedNodes failed: ${e.message}")
                false
            }
            mainHandler.post { callback(reachable) }
        }
    }

    // -- Session commands -----------------------------------------------------

    /**
     * Send a start-session command to the watch.
     *
     * @param config  JSON map of session configuration.
     * @param callback Called for each event the watch streams back.
     */
    fun startSession(config: Map<String, Any?>, callback: (Map<String, Any?>) -> Unit) {
        activeSessionId = config["session_id"] as? String
        eventCallback = callback

        val message = JSONObject().apply {
            put("command", "start_session")
            put("session_id", config["session_id"])
            put("mode", config["mode"])
            put("duration_sec", config["duration_sec"])
            val profile = config["profile"]
            if (profile is Map<*, *>) {
                put("profile", JSONObject().apply {
                    put("window_sec", profile["window_sec"])
                    put("emit_interval_sec", profile["emit_interval_sec"])
                })
            }
            val label = config["window_label"]
            if (label != null) put("window_label", label)
        }

        val payload = message.toString().toByteArray(Charsets.UTF_8)

        scope.launch {
            try {
                val nodes = nodeClient.connectedNodes.await()
                for (node in nodes) {
                    try {
                        messageClient.sendMessage(node.id, COMMAND_PATH, payload).await()
                    } catch (e: Exception) {
                        Log.w(TAG, "sendMessage failed to ${node.displayName}: ${e.message}")
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "startSession: no connected nodes: ${e.message}")
                mainHandler.post { cleanup() }
            }
        }
    }

    /** Send a stop-session command to the watch. */
    fun stopSession(sessionId: String) {
        val message = JSONObject().apply {
            put("command", "stop_session")
            put("session_id", sessionId)
        }
        val payload = message.toString().toByteArray(Charsets.UTF_8)

        scope.launch {
            try {
                val nodes = nodeClient.connectedNodes.await()
                for (node in nodes) {
                    try {
                        messageClient.sendMessage(node.id, COMMAND_PATH, payload).await()
                    } catch (e: Exception) {
                        Log.w(TAG, "stop sendMessage failed to ${node.displayName}: ${e.message}")
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "stopSession failed: ${e.message}")
            }
        }
    }

    // -- Receiving events from watch ------------------------------------------

    override fun onMessageReceived(messageEvent: MessageEvent) {
        if (messageEvent.path != EVENT_PATH) return

        val json = JSONObject(String(messageEvent.data, Charsets.UTF_8))
        val eventMap = jsonToMap(json)
        val type = eventMap["type"] as? String ?: ""
        Log.d(TAG, "onMessageReceived: type=$type path=${messageEvent.path} sourceNode=${messageEvent.sourceNodeId}")

        mainHandler.post {
            if (eventCallback == null) {
                Log.w(TAG, "onMessageReceived: eventCallback is null, dropping event type=$type")
            } else {
                eventCallback?.invoke(eventMap)
            }

            if (type == "session_summary" || type == "session_error") {
                cleanup()
            }
        }
    }

    // -- Lifecycle ------------------------------------------------------------

    fun dispose() {
        messageClient.removeListener(this)
        cleanup()
        scope.cancel()
    }

    private fun cleanup() {
        activeSessionId = null
        eventCallback = null
    }

    // -- Helpers --------------------------------------------------------------

    private fun isPlayServicesAvailable(): Boolean {
        return GoogleApiAvailability.getInstance()
            .isGooglePlayServicesAvailable(context) == ConnectionResult.SUCCESS
    }

    /** Convert JSONObject to Map<String, Any?> recursively. */
    private fun jsonToMap(json: JSONObject): Map<String, Any?> {
        val map = mutableMapOf<String, Any?>()
        val keys = json.keys()
        while (keys.hasNext()) {
            val key = keys.next()
            val value = json.get(key)
            map[key] = when (value) {
                is JSONObject -> jsonToMap(value)
                is org.json.JSONArray -> jsonArrayToList(value)
                JSONObject.NULL -> null
                else -> value
            }
        }
        return map
    }

    private fun jsonArrayToList(array: org.json.JSONArray): List<Any?> {
        return (0 until array.length()).map { i ->
            when (val value = array.get(i)) {
                is JSONObject -> jsonToMap(value)
                is org.json.JSONArray -> jsonArrayToList(value)
                JSONObject.NULL -> null
                else -> value
            }
        }
    }
}

File: ios/Classes/WatchSessionRelay.swift
import Foundation
import WatchConnectivity

/// Phone-side relay that bridges session commands and HSI events between the
/// Flutter plugin and a companion watchOS app via `WCSession`.
///
/// Lifecycle:
///   1. Plugin calls `startSession` → relay sends command to watch.
///   2. Watch engine emits events → relay receives them in `didReceiveMessage`
///      and forwards to the stored `eventCallback`.
///   3. On terminal events (`session_summary` / `error`) the relay cleans up
///      automatically.
///
/// If `sendMessage` fails (watch unreachable, app not installed, etc.) the
/// `onSendFailed` closure fires so the plugin can fall back to a local engine.
class WatchSessionRelay: NSObject, WCSessionDelegate {

    typealias EventCallback = ([String: Any]) -> Void

    private var eventCallback: EventCallback?
    private var activeSessionId: String?
    private var activated = false

    override init() {
        super.init()
        guard WCSession.isSupported() else { return }
        WCSession.default.delegate = self
        WCSession.default.activate()
    }

    // MARK: - Status

    /// Full connectivity snapshot for the Dart layer.
    var status: [String: Any] {
        guard WCSession.isSupported(), activated else {
            return ["supported": WCSession.isSupported(), "reachable": false,
                    "paired": false, "installed": false]
        }
        let s = WCSession.default
        return ["supported": true, "reachable": s.isReachable,
                "paired": s.isPaired, "installed": s.isWatchAppInstalled]
    }

    /// Whether the watch companion is reachable right now.
    /// This is the *authoritative* check — `isReachable` requires the watch app
    /// to be installed, paired, and in the foreground.
    var isReachable: Bool {
        guard WCSession.isSupported(), activated else { return false }
        return WCSession.default.isReachable
    }

    // MARK: - Session commands

    /// Send a start-session command to the watch.
    ///
    /// - Parameters:
    ///   - config: Session configuration to forward.
    ///   - callback: Called for each event the watch streams back.
    ///   - onSendFailed: Called (on main queue) if `sendMessage` fails,
    ///     allowing the caller to fall back to a local engine.
    func startSession(config: SessionConfig,
                      callback: @escaping EventCallback,
                      onSendFailed: @escaping () -> Void) {
        activeSessionId = config.sessionId
        eventCallback = callback

        var message: [String: Any] = [
            "command": "start_session",
            "session_id": config.sessionId,
            "mode": config.mode.rawValue,
            "duration_sec": config.durationSec,
            "profile": [
                "window_sec": config.profile.windowSec,
                "emit_interval_sec": config.profile.emitIntervalSec
            ]
        ]
        if let label = config.windowLabel {
            message["window_label"] = label
        }

        WCSession.default.sendMessage(message, replyHandler: nil) { [weak self] error in
            NSLog("[WatchRelay] sendMessage failed: \(error.localizedDescription)")
            DispatchQueue.main.async {
                self?.cleanup()
                onSendFailed()
            }
        }
    }

    /// Send a stop-session command to the watch.
    func stopSession(sessionId: String) {
        WCSession.default.sendMessage(
            ["command": "stop_session", "session_id": sessionId],
            replyHandler: nil
        ) { error in
            NSLog("[WatchRelay] stop sendMessage failed: \(error.localizedDescription)")
        }
    }

    // MARK: - WCSessionDelegate (iOS required)

    func session(_ session: WCSession,
                 activationDidCompleteWith activationState: WCSessionActivationState,
                 error: Error?) {
        activated = (activationState == .activated)
        if let error = error {
            NSLog("[WatchRelay] activation error: \(error.localizedDescription)")
        }
    }

    func sessionDidBecomeInactive(_ session: WCSession) {}

    func sessionDidDeactivate(_ session: WCSession) {
        WCSession.default.activate()
    }

    // MARK: - Receiving events from watch

    func session(_ session: WCSession, didReceiveMessage message: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.eventCallback?(message)

            let type = message["type"] as? String ?? ""
            if type == "session_summary" || type == "session_error" {
                self?.cleanup()
            }
        }
    }

    // MARK: - Private

    private func cleanup() {
        activeSessionId = nil
        eventCallback = nil
    }
}
